codebleu/test_codebleu.py

```diff
@@ -0,0 +1,180 @@
+import json
+import unittest
+import collections
+import codebleu_metric as cbl
+import codebleu.graph_generator.graphgenerator as gg
+import networkx as nx
+from graph_generator.graphgenerator import AstGraphGenerator
+from graph_generator.graphgenutils import prettyprint_graph
+from graph_generator.type_lattice_generator import TypeLatticeGenerator
+from tree_sitter import Language, Parser
+
+
+class TestCodebleuMetric(unittest.TestCase):
+    # ---- DFG part ----
+    parser = Parser()
+    PY_LANGUAGE = Language('codebleu/my-languages.so', 'python')
+    parser.set_language(PY_LANGUAGE)
+    lattice = TypeLatticeGenerator('codebleu/typingRules.json')
+
+    def test_inlist(self):
+        self.assertEqual(cbl._inlist(1, [1, 2, 3]), True)
+        self.assertEqual(cbl._inlist(1, [5, 2, 3]), False)
+        self.assertEqual(cbl._inlist(1, 1), True)
+        self.assertEqual(cbl._inlist(1, 2), False)
+
+    def test_split_identifier(self):
+        self.assertEqual(cbl._split_identifier('1_asd'), [1, 'asd'])
+        self.assertEqual(cbl._split_identifier('1_asd_e'), [1, 'asd_e'])
+        self.assertRaises(Exception, cbl._split_identifier('1_'))
+        self.assertRaises(Exception, cbl._split_identifier('123'))
+        self.assertRaises(Exception, cbl._split_identifier('asd_a'))
+        self.assertRaises(Exception, cbl._split_identifier('___'))
+        self.assertRaises(Exception, cbl._split_identifier('_asd'))
```

Comment:
Here you can also use parameterized tests. In order to use exceptions in them, you will need one more trick:
```suggestion
    @pytest.mark.parametrize(
        "identifier, outcome, expectation",
        [
            ("1_asd", [1, "asd"], does_not_raise()),
            ("1_asd_e", [1, "asd_e"], does_not_raise()),
            ("1_", None, pytest.raises(Exception)),
            ("123", None, pytest.raises(Exception)),
            ("asd_a", None, pytest.raises(Exception)),
            ("___", None, pytest.raises(Exception)),
            ("_asd", None, pytest.raises(Exception)),
        ],
    )
    def test_split_identifier(self, identifier, outcome, expectation):
        with expectation:
            assert cbl._split_identifier(identifier) == outcome
```

Also, add `from contextlib import nullcontext as does_not_raise` to imports. This way you will also find out that some of the calls to `_split_identifier` do not raise Exceptions :(

Reply:
Also, `Exception` is very generic, can we use something more specific here?